DebugAdapterExecutable does not support passing `shell: true` so `.bat`-based debug adapters are broken in current Insiders

In current Insiders (and presumably the upcoming release), spawning a `.bat` file on Windows without `shell: true` fails due to a breaking change in NodeJS:

https://nodejs.org/en/blog/vulnerability/april-2024-security-releases-2

> It is important to note that there has been a breaking change for Windows users who utilize `child_process.spawn` and `child_process.spawnSync`. Node.js will now error with `EINVAL` if a `.bat` or `.cmd` file is passed to `child_process.spawn` and `child_process.spawnSync` without the `shell` option set.

It appears that there's no way to signal this to [DebugAdapterExecutable](https://code.visualstudio.com/api/references/vscode-api#DebugAdapterExecutable) and therefore any `.bat`-based debug adapters launched this way fail in current Insiders (and will presumably fail in stable soon).

Could `shell: bool?` be added to `DebugAdapterExecutableOptions`? I don't know enough to contribute this, but this seems quite important.

I started trying to fix this here:

https://github.com/DanTup/vscode/commit/750a66d8f62a59fb027ea487c2484f0d749c63e2

However I'm unable to get VS Code compiling locally, it spews lots of errors that don't make much sense to me:

```
[watch-extensions] [14:54:13] Error: C:/Dev/Microsoft/vscode/.vscode/extensions/vscode-selfhost-test-provider/src/importGraph.ts(8,26): Cannot find module 'cockatiel' or its corresponding type declarations.
[watch-extensions] [14:54:13] Error: C:/Dev/Microsoft/vscode/extensions/css-language-features/client/src/browser/cssClientMain.ts(22,40): Type 'Worker' has no properties in common with type 'LanguageClientOptions'.
[watch-extensions] [14:54:13] Error: C:/Dev/Microsoft/vscode/extensions/html-language-features/client/src/browser/htmlClientMain.ts(21,40): Type 'Worker' has no properties in common with type 'LanguageClientOptions'.
[watch-extensions] [14:54:13] Error: C:/Dev/Microsoft/vscode/extensions/git/src/encoding.ts(83,83): Argument of type '{ detectEncodings: string[]; } | undefined' is not assignable to parameter of type '{ minimumThreshold: number; } | undefined'.
[watch-extensions]   Object literal may only specify known properties, and 'detectEncodings' does not exist in type '{ minimumThreshold: number; }'.
[watch-extensions] [14:54:13] Error: C:/Dev/Microsoft/vscode/extensions/json-language-features/client/src/browser/jsonClientMain.ts(21,40): Type 'Worker' has no properties in common with type 'LanguageClientOptions'.
[watch-extensions] [14:54:13] Error: C:/Dev/Microsoft/vscode/extensions/vscode-api-tests/src/singlefolder-tests/proxy.test.ts(9,21): Cannot find module 'node-forge' or its corresponding type declarations.
[watch-extensions] [14:54:13] Error: C:/Dev/Microsoft/vscode/extensions/vscode-api-tests/src/singlefolder-tests/proxy.test.ts(13,45): Cannot find module 'straightforward' or its corresponding type declarations.
[watch-extensions] [14:54:13] Error: C:/Dev/Microsoft/vscode/extensions/vscode-api-tests/src/singlefolder-tests/proxy.test.ts(88,27): Parameter 'context' implicitly has an 'any' type.
[watch-extensions] [14:54:13] Error: C:/Dev/Microsoft/vscode/extensions/vscode-api-tests/src/singlefolder-tests/proxy.test.ts(88,36): Parameter 'next' implicitly has an 'any' type.
[watch-extensions] [14:54:13] Error: C:/Dev/Microsoft/vscode/extensions/vscode-api-tests/src/singlefolder-tests/proxy.test.ts(94,23): Binding element 'clientSocket' implicitly has an 'any' type.
[watch-extensions] [14:54:13] Error: C:/Dev/Microsoft/vscode/extensions/typescript-language-features/src/languageFeatures/fileConfigurationManager.ts(194,4): Object literal may only specify known properties, and 'preferTypeOnlyAutoImports' does not exist in type 'UserPreferences'.
[watch-extensions] [14:54:13] Error: C:/Dev/Microsoft/vscode/extensions/typescript-language-features/src/typescriptService.ts(79,20): 'ts.server.protocol' has no exported member named 'MapCodeRequestArgs'. Did you mean 'CodeFixRequestArgs'?
[watch-extensions] [14:54:13] Error: C:/Dev/Microsoft/vscode/extensions/typescript-language-features/src/typescriptService.ts(79,46): 'ts.server.protocol' has no exported member named 'MapCodeResponse'. Did you mean 'CodeFixResponse'?
[watch-extensions] [14:54:13] Error: C:/Dev/Microsoft/vscode/extensions/typescript-language-features/src/typescriptService.ts(80,26): Namespace 'ts.server.protocol' has no exported member 'GetPasteEditsRequestArgs'.
[watch-extensions] [14:54:13] Error: C:/Dev/Microsoft/vscode/extensions/typescript-language-features/src/typescriptService.ts(80,58): Namespace 'ts.server.protocol' has no exported member 'GetPasteEditsResponse'.
```

I'm happy for someone else to fix this if it's faster :-)

--- a/src/vs/workbench/contrib/debug/node/debugAdapter.ts
+++ b/src/vs/workbench/contrib/debug/node/debugAdapter.ts
@@ -223,6 +223,7 @@ export class ExecutableDebugAdapter extends StreamDebugAdapter {
 				}
 			} else {
 				let spawnCommand = command;
+				let spawnArgs = args;
 				const spawnOptions: cp.SpawnOptions = {
 					env: env
 				};
@@ -233,9 +234,10 @@ export class ExecutableDebugAdapter extends StreamDebugAdapter {
 					// https://github.com/microsoft/vscode/issues/224184
 					spawnOptions.shell = true;
 					spawnCommand = `"${command}"`;
+					spawnArgs = args.map(a => `"${a}"`);
 				}
 
-				this.serverProcess = cp.spawn(spawnCommand, args, spawnOptions);
+				this.serverProcess = cp.spawn(spawnCommand, spawnArgs, spawnOptions);
 			}
 
 			this.serverProcess.on('error', err => {

--- a/src/vs/workbench/contrib/debug/node/debugAdapter.ts
+++ b/src/vs/workbench/contrib/debug/node/debugAdapter.ts
@@ -230,7 +230,7 @@ export class ExecutableDebugAdapter extends StreamDebugAdapter {
 				if (options.cwd) {
 					spawnOptions.cwd = options.cwd;
 				}
-				if (platform.isWindows && command.endsWith('.bat')) {
+				if (platform.isWindows && (command.endsWith('.bat') || command.endsWith('.cmd'))) {
 					// https://github.com/microsoft/vscode/issues/224184
 					spawnOptions.shell = true;
 					spawnCommand = `"${command}"`;

--- a/src/vs/workbench/contrib/debug/node/debugAdapter.ts
+++ b/src/vs/workbench/contrib/debug/node/debugAdapter.ts
@@ -235,7 +235,6 @@ export class ExecutableDebugAdapter extends StreamDebugAdapter {
 					spawnOptions.shell = true;
 					spawnCommand = `"${command}"`;
 					spawnArgs = args.map(a => {
-						a = a.replace(/\\/g, '\\\\'); // Escape existing \
 						a = a.replace(/"/g, '\\"'); // Escape existing double quotes with \
 						// Wrap in double quotes
 						return `"${a}"`;

--- a/src/vs/workbench/contrib/debug/node/debugAdapter.ts
+++ b/src/vs/workbench/contrib/debug/node/debugAdapter.ts
@@ -222,13 +222,20 @@ export class ExecutableDebugAdapter extends StreamDebugAdapter {
 					throw new Error(nls.localize('unableToLaunchDebugAdapterNoArgs', "Unable to launch debug adapter."));
 				}
 			} else {
+				let spawnCommand = command;
 				const spawnOptions: cp.SpawnOptions = {
 					env: env
 				};
 				if (options.cwd) {
 					spawnOptions.cwd = options.cwd;
 				}
-				this.serverProcess = cp.spawn(command, args, spawnOptions);
+				if (platform.isWindows && command.endsWith('.bat')) {
+					// https://github.com/microsoft/vscode/issues/224184
+					spawnOptions.shell = true;
+					spawnCommand = `"${command}"`;
+				}
+
+				this.serverProcess = cp.spawn(spawnCommand, args, spawnOptions);
 			}
 
 			this.serverProcess.on('error', err => {

--- a/src/vs/workbench/contrib/debug/node/debugAdapter.ts
+++ b/src/vs/workbench/contrib/debug/node/debugAdapter.ts
@@ -234,7 +234,12 @@ export class ExecutableDebugAdapter extends StreamDebugAdapter {
 					// https://github.com/microsoft/vscode/issues/224184
 					spawnOptions.shell = true;
 					spawnCommand = `"${command}"`;
-					spawnArgs = args.map(a => `"${a}"`);
+					spawnArgs = args.map(a => {
+						a = a.replace(/\\/g, '\\\\'); // Escape existing \
+						a = a.replace(/"/g, '\\"'); // Escape existing double quotes with \
+						// Wrap in double quotes
+						return `"${a}"`;
+					});
 				}
 
 				this.serverProcess = cp.spawn(spawnCommand, spawnArgs, spawnOptions);

